cpp/src/arrow/chunk_resolver.h

```diff
@@ -0,0 +1,93 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <atomic>
+#include <cstdint>
+#include <vector>
+
+#include "arrow/type_fwd.h"
+#include "arrow/util/macros.h"
+
+namespace arrow {
+namespace internal {
+
+struct ChunkLocation {
+  int64_t chunk_index, index_in_chunk;
+};
+
+// An object that resolves an array chunk depending on the index
+struct ChunkResolver {
+  ChunkResolver(const ChunkResolver& chunks);
```

Comment:
If a copy constructor is needed, can we add a move constructor as well?

Reply:
A move constructor is not allowed bc it has a `std::atomic` member variable which do not support copy/move operations.

The copy constructor is only needed bc [`TableSorter` has a `ChunkResolver` member](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/vector_sort.cc#L1140) and performs copies. Actually, I think the copies are from [`ResolvedSortKey` and bc it is declared inside the class](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/vector_sort.cc#L843), the copy/move rules are applied based on the enclosing class. [Here is a copy operation for `ResolvedSortKey`.](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/vector_sort.cc#L547)

Reply:
I am checking this in more details.

Reply:
I think it would be better to identify and apply a workaround the location(s) where `ChunkResolver` is being copied/moved, then we can eliminate the copy constructor from `ChunkResolver`.

Reply:
I think the copy operation of `ResolvedSortKey` occurs in [this definition in `TableSelecter` that contains a `ChunkedArrayResolver` member](https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/vector_sort.cc#L1692) (which inherits from `ChunkResolver`).

Reply:
I moved the copy constructor to `ChunkedArrayResolver` in order to support copy operations of `ResolvedSortKey` from `TableSelecter`.